File: src/services/api/engine.ts
import { apiRequest } from '@/lib/api-utils';

const API_BASE = '/api/engine';

export interface EngineConfig {
  id?: number;
  user_id: string;
  selected_model: string;
  temperature: number;
  max_tokens: number;
  vibe: string;
  audience: string;
  updated_at?: string;
}

export interface TelemetryData {
  model: string;
  latency_ms: number;
  status: 'SUCCESS' | 'ERROR' | 'FALLBACK';
  endpoint: string;
  request_summary?: string;
  error_message?: string;
  metadata?: any;
}

export const engineApi = {
  getConfig: async (userId: string): Promise<EngineConfig> => {
    return apiRequest<EngineConfig>(`${API_BASE}/config/${userId}`, {
      method: 'GET'
    });
  },

  updateConfig: async (userId: string, update: Partial<EngineConfig>): Promise<EngineConfig> => {
    return apiRequest<EngineConfig>(`${API_BASE}/config/${userId}`, {
      method: 'POST',
      body: JSON.stringify(update)
    });
  },

  recordTelemetry: async (data: TelemetryData, userId?: string): Promise<void> => {
    await apiRequest(`${API_BASE}/telemetry${userId ? `?user_id=${userId}` : ''}`, {
      method: 'POST',
      body: JSON.stringify(data)
    });
  },

  getRecentTelemetry: async (limit: number = 50): Promise<any[]> => {
    return apiRequest<any[]>(`${API_BASE}/telemetry/recent?limit=${limit}`, {
      method: 'GET'
    });
  }
};

File: backend/api/engine.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import select
from sqlmodel.ext.asyncio.session import AsyncSession
from typing import List, Optional
from datetime import datetime
from backend.database import get_async_session
from backend.models.engine import EngineConfig, AITelemetry
from pydantic import BaseModel

router = APIRouter(prefix="/api/engine", tags=["Engine Nexus"])

class EngineConfigUpdate(BaseModel):
    selected_model: Optional[str] = None
    temperature: Optional[float] = None
    max_tokens: Optional[int] = None
    vibe: Optional[str] = None
    audience: Optional[str] = None

class TelemetryCreate(BaseModel):
    model: str
    latency_ms: float
    status: str = "SUCCESS"
    endpoint: str
    request_summary: Optional[str] = None
    error_message: Optional[str] = None
    metadata: dict = {}

@router.get("/config/{user_id}", response_model=EngineConfig)
async def get_engine_config(user_id: str, session: AsyncSession = Depends(get_async_session)):
    statement = select(EngineConfig).where(EngineConfig.user_id == user_id)
    result = await session.exec(statement)
    config = result.first()
    
    if not config:
        # Create default config for new users
        config = EngineConfig(user_id=user_id)
        session.add(config)
        await session.commit()
        await session.refresh(config)
    
    return config

@router.post("/config/{user_id}", response_model=EngineConfig)
async def update_engine_config(user_id: str, update: EngineConfigUpdate, session: AsyncSession = Depends(get_async_session)):
    statement = select(EngineConfig).where(EngineConfig.user_id == user_id)
    result = await session.exec(statement)
    config = result.first()
    
    if not config:
        config = EngineConfig(user_id=user_id)
        session.add(config)

    for key, value in update.dict(exclude_unset=True).items():
        setattr(config, key, value)
    
    config.updated_at = datetime.utcnow()
    session.add(config)
    await session.commit()
    await session.refresh(config)
    return config

@router.post("/telemetry", status_code=201)
async def record_telemetry(telemetry: TelemetryCreate, user_id: Optional[str] = None, session: AsyncSession = Depends(get_async_session)):
    db_telemetry = AITelemetry(
        user_id=user_id,
        model=telemetry.model,
        latency_ms=telemetry.latency_ms,
        status=telemetry.status,
        endpoint=telemetry.endpoint,
        request_summary=telemetry.request_summary,
        error_message=telemetry.error_message,
        extra_metadata=telemetry.metadata
    )
    session.add(db_telemetry)
    await session.commit()
    return {"status": "recorded"}

@router.get("/telemetry/recent", response_model=List[AITelemetry])
async def get_recent_telemetry(limit: int = 50, session: AsyncSession = Depends(get_async_session)):
    statement = select(AITelemetry).order_by(AITelemetry.timestamp.desc()).limit(limit)
    result = await session.exec(statement)
    return result.all()
